### PR TITLE
Added use-onchip-memory sample

### PR DESCRIPTION
**Pull request by cjdb**

### CODE REVIEW

**Comment by MathiasMagnus** [comment]
This sounds awesome, but what on Earth is the `on_chip_memory` extension? :D If you'd have a link in the comments, I'd definitely read it before scanning through the sample.

**Comment by keryell** [comment]
Yes the example should direct somehow to a specification.
@MathiasMagnus full disclosure is on https://github.com/codeplaysoftware/standards-proposals/pull/83
I am pretty sure you would love following this git repository. :-)
But by the way, when do you join the Khronos SYCL group? :)

**Comment by MathiasMagnus** [comment]
I'm an unfrequent committer to this repo, but will star it. ;) As for joining Khronos… next year's funding is _problematic_ to say the least. I'm willing to discuss in greater detail, but not on public forums. :)

File: samples/use-onchip-memory/use-onchip-memory.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  auto queue = sycl::queue{};
+  ::use_with_policy(codeplay::property::prefer, queue);
+}
+
+// Alternatively, if you can guarantee that your system will support this
+// policy, or if it is expected any system using your software must support the
+// policy, then you can use the require tag to indicate that the feature is
+// required by your software.
+//
+// In the event that the property isn't supported, an exception will be thrown.
```
Comment (Ruyk) [inline samples/use-onchip-memory/use-onchip-memory.cpp]: Which exception? is there a particular SYCL exception thrown?

File: samples/use-onchip-memory/use-onchip-memory.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+#include <iostream>
+
+namespace sycl = cl::sycl;
+namespace access = sycl::access;
+
+// Sample kernel
+// -------------
+// Since kernels are no longer allowed to be declared inline, I dislike
+// declaring a random name in the global namespace that can potentially be
+// repurposed.
```
[the remark this answers is not in the document]
Comment (DuncanMcBain) [reply samples/use-onchip-memory/use-onchip-memory.cpp]: I'd go further than that and say you don't need to justify yourself :) some samples have lambdas, some have function objects, but I don't think it needs commentary.

File: samples/use-onchip-memory/use-onchip-memory.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+// This kernel performs the same operation as std::iota, but also scales
+// the result by two.
+//
+class scaled_iota {
+ public:
+  using accessor = sycl::accessor<int, 1, access::mode::discard_write,
+                                  access::target::global_buffer>;
+
+  explicit scaled_iota(accessor deviceAccess) noexcept
+      : m_deviceAccess{std::move(deviceAccess)} {}
```
Comment (DuncanMcBain) [inline samples/use-onchip-memory/use-onchip-memory.cpp]: I don't think you get any benefit by moving accessors

File: samples/use-onchip-memory/use-onchip-memory.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    auto deviceData = sycl::buffer<int, 1>{
+        hostData.data(),                  //
+        sycl::range<1>(hostData.size()),  //
+        sycl::property_list{
+            sycl::property::buffer::context_bound(taskContext),
+            codeplay::property::buffer::use_onchip_memory(policy)  // <--------
+        }  //                                            ^---------------------
+    };
+
+    deviceData.set_final_data(hostData.data());
```
Comment (DuncanMcBain) [inline samples/use-onchip-memory/use-onchip-memory.cpp]: I don't think you need this if you also set hostData.data() as the first argument to the deviceData ctor (or are you thinking of doing a device-only buffer instead?)

File: samples/use-onchip-memory/use-onchip-memory.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+// feature is not present on the system, then it will not be enabled.
+//
+// Puns aside, this is the preferred default.
+//
+void how_to_use_with_prefer() {
+  auto queue = sycl::queue{};
+  ::use_with_policy(codeplay::property::prefer, queue);
+}
+
+// Alternatively, if you can guarantee that your system will support this
```
Comment (DuncanMcBain) [inline samples/use-onchip-memory/use-onchip-memory.cpp]: You could also mention here that making it required lets your program know if it is available or not. Preferred will choose one but there's no way to know which. (I don't know if that's useful, maybe you tweak the algorithm slightly if onchip is available or something)

File: samples/use-onchip-memory/use-onchip-memory.cpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  accessor m_deviceAccess;
+};
+
+namespace codeplay = sycl::codeplay;
+
+template <class Policy>
+void use_with_policy(Policy policy, sycl::queue& queue) {
+  auto hostData = sycl::vector_class<int>(1024);
+  {
+    auto taskContext = queue.get_context();
```
[the remark this answers is not in the document]
Comment (cjdb) [reply samples/use-onchip-memory/use-onchip-memory.cpp]: What else could it be named as? I'm not a fan of naming a variable its type.

File: samples/use-onchip-memory/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,21 @@
+set(SOURCE_NAME "use-onchip-memory")
+
+add_executable(
+   ${SOURCE_NAME}
+   ${CMAKE_CURRENT_SOURCE_DIR}/${SOURCE_NAME}.cpp
+)
+include_directories(
```
[the remark this answers is not in the document]
Comment (cjdb) [reply samples/use-onchip-memory/CMakeLists.txt]: Removing it didn't seem to affect compilation.

File: samples/use-onchip-memory/use-onchip-memory.cpp (hunk 5 above)
Comment (cjdb) [reply samples/use-onchip-memory/use-onchip-memory.cpp]: Isn't that what I've said?

Comment (DuncanMcBain) [reply samples/use-onchip-memory/use-onchip-memory.cpp]: Yes, sort of. I guess the consequence flows naturally, and is a slightly more unusual use-case.

File: samples/use-onchip-memory/use-onchip-memory.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ruyk) [reply samples/use-onchip-memory/use-onchip-memory.cpp]: executionContext is probably the name would make you the happiest

File: samples/use-onchip-memory/use-onchip-memory.cpp [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+void how_to_use_with_require(sycl::queue& queue) {
+  try {
+    ::use_with_policy(codeplay::property::require, queue);
+  } catch (const sycl::exception& e) {
+    std::cerr << "An error occurred: " << e.what()
+              << "\n"
+                 "\n"
+                 "This particular error has occurred because you are requiring "
+                 "the policy use_onchip_memory be available, and your hardware "
+                 "doesn't support the use_onchip_memory, so the SYCL ecosystem "
```
Comment (DuncanMcBain) [inline samples/use-onchip-memory/use-onchip-memory.cpp]: Might be clearer to say "implementation" instead of "ecosystem".

Comment (cjdb) [reply samples/use-onchip-memory/use-onchip-memory.cpp]: Agreed.

**Pull request merged**